pkg/apis/azuredisk/v1alpha1/types.go

```diff
@@ -56,24 +57,50 @@ type AzVolumeList struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // AzVolumeAttachment is a specification for a AzVolumeAttachment resource
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="NodeName",type=string,JSONPath=`.spec.nodeName`,description="Name of the Node which this AzVolumeAttachment object is attached to"
+// +kubebuilder:printcolumn:name="UnderlyingVolume",type=string,JSONPath=`.spec.underlyingVolume`,description="Name of the Volume which this AzVolumeAttachment object references"
+// +kubebuilder:printcolumn:name="Role",type=string,JSONPath=`.spec.role`,description="Indicates if the volume attachment should be primary attachment or not"
+// +kubebuilder:printcolumn:name="Role",type=string,JSONPath=`.status.role`,description="Indicates if the volume attachment is primary attachment or not"
 type AzVolumeAttachment struct {
-	metav1.TypeMeta   `json:",inline"`
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
-	Spec   AzVolumeAttachmentSpec   `json:"spec"`
-	Status AzVolumeAttachmentStatus `json:"status"`
+	// spec defines the desired state of a AzVolumeAttachment.
+	// Required.
+	Spec AzVolumeAttachmentSpec `json:"spec"`
+
+	// status represents the current state of AzVolumeAttachment.
+	// Nil status indicates that the underlying volume of AzVolumeAttachment has not yet been attached to the specified node
+	// +optional
+	Status *AzVolumeAttachmentStatus `json:"status,omitempty"`
 }
 
 // AzVolumeAttachmentSpec is the spec for a AzVolumeAttachment resource
 type AzVolumeAttachmentSpec struct {
 	UnderlyingVolume string `json:"underlyingVolume"`
-	NodeName         string `json:"NodeName"`
-	Partition        int32  `json:"partition"`
+	NodeName         string `json:"nodeName"`
+	Role             Role   `json:"role"`
 }
 
+// Role indicates if the volume attachment is replica attachment or not
+type Role string
+
+const (
+	// Primary indicates that the specified node the volume is attached to is where the pod is currently running on
+	Primary Role = "Primary"
+	// Replica indicates that the specified node the volume is attached to is one of replicas that pod can failover to if the primary node fails
+	Replica Role = "Replica"
+)
+
 // AzVolumeAttachmentStatus is the status for a AzVolumeAttachment resource
 type AzVolumeAttachmentStatus struct {
-	State string `json:"state,omitempty"`
+	Role Role `json:"role"`
 }
```

Comment:
Do we need to keep the `Role` both in the spec and status of the `AzVolumeAttachment`? I think it should suffice to have the `Role` only in `AzVolumeAttachmentSpec` and `AzVolumeAttachmentStatus` should be left as if for reflecting the state of the object (i.e. for example if we fail to delete the volumeAttachment,  the state should reflect this error so that other controllers and the scheduler can rely on this info).

Reply:
`Role` in `AzVolumeAttachmentStatus` refers to the current `Role` the attachment is assigned and `Role` in `AzVolumeAttachmentSepc` refers to the desired `Role`. This will be used in replica control mainly as pod failover means a transition of role from `Replica` to `Primary` for an attachment. `ControllerPublishVolume` will be updating the spec `Role` of a replica the pod will be failing over to. When the controller sees such update and a discrepancy between `Role` in spec and status, it will update the role in status and create another replica to replace the replica that has been promoted to `Primary` role.

Any volumeAttachment that should be deleted will have a deletion timestamp set, so hopefully scheduler extender can rely on that. If we need more information on volumeAttachment object going forward, we can add an additional status field in future PR.

Reply:
Makes sense. Consider naming it `RequestedRole` in the `Spec` to make this clearer.

Reply:
Sounds good. Will do :)

Reply:
Ah, I see, this is the spec and status convention in Kubernetes, I agree that renaming it to `RequestedRole ` will make it more clear. I still think that the controller should have and report the state of the object itself. The fact that after deletion, the volumeAttachment will have a deletion timestamp is an implementation detail that other resources should not know or rely on. Also, deletion is not the only case here - we could have a failure on create/startup and the status field should reflect when the resource is ready to be used. See the description of status properties [here ](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)

Reply:
With the current implementation, the status field is a pointer to a `AzVolumeAttachmentStatus` object. So any `AzVolumeAttachment` object that is not in a ready state will have a `nil` status in order to keep things as simple as possible. And once volume attachment is complete, the status field will be initialized to indicate the object's readiness. Moreover, `AzVolumeAttachment` in a ready state should not have any events to update its status other than when a replica object is being promoted to a primary role.

One concern I have for adding an additional condition status field under `azvolumeattachment.status` is that it will inevitably require more http calls, as `azvolumeattachment.status` is a subresource that requires a separate update call from the main resource update call, which can add more point of failures.

I understand your point though. Before we proceed, I think it would be helpful to think of specific use cases that would require a condition status field. :)

Reply:
Ah, if there might be a performance impact and there are no intermediate states/phases for the attachment, I agree that it would probably be ok to just rely on the `Role`. This would make the resource conceptually similar to replica sets in terms of how it reports its state. I had a mental model that it should be similar to persistent volumes/pods/nodes...